File: all/2215_find_the_difference_of_two_arrays/sol_sorting_deduplicate_inplace_two_pointers.py
from typing import List


def deduplicateSortedList(sorted_list: List[int]) -> None:
    next_unique_element_index = 1

    for i in range(1, len(sorted_list)):
        if (sorted_list[i - 1]) == (sorted_list[i]):
            # Skip repeated values.
            continue

        # Move the current found unique value to the position that
        # the next_unique_element_index points to.
        sorted_list[next_unique_element_index] = (sorted_list[i])
        next_unique_element_index += 1

    # Only preserve the part with all the unique values.
    del sorted_list[next_unique_element_index:]

class Solution:
    def findDifference(
        self,
        nums1: List[int],
        nums2: List[int]
    ) -> List[List[int]]:
        answer_list = list()

        # Sort the nums1 list in the ascending order.
        nums1.sort()

        # Sort the nums2 list in the ascending order.
        nums2.sort()

        # Deduplicate the nums1 list inplace.
        deduplicateSortedList(nums1)

        # Deduplicate the nums2 list inplace.
        deduplicateSortedList(nums2)

        def getDifferenceBetweenTwoIntListsValues(
            list_1: List[int],
            list_2: List[int]
        ) -> List[int]:
            # Assume list 1 and list 2 are both sorted in the ascending order
            # and both lists have unique values.
            #
            # We use two pointers to find all unique values that are the list 1
            # but not in the list 2.
            i = 0
            j = 0
            values_list = list()

            while (i < len(list_1)) and (j < len(list_2)):
                # As list_1[i] is equal to list_2[j], it cannot be the value
                # we want, just increase the index i and the index j by 1.
                # The index j is also increased since all values in the list 1
                # and the list 2 are unique. There won't be other elements
                # whose value is list_1[i] in the rest of list 1.
                if (list_1[i]) == (list_2[j]):
                    i += 1
                    j += 1
                elif (list_1[i]) > (list_2[j]):
                    # As list_1[i] is greater than list_2[j], we do not know
                    # whether list_1[i] only shows in the int array 1, we need
                    # to increase the index j by 1 to check more values.
                    j += 1
                else:
                    # As list_1[i] is less than list_2[j], and since list 2 is
                    # sorted, we cannot find list_1[i] in the list_2[j:],
                    # so list_1[i] must only show in the list 1.
                    # Also, all values in the given int array 1 are unique,
                    # we can add list_1[i] to answer array.
                    values_list.append(list_1[i])

                    i += 1

            while (i < len(list_1)):
                # If we still not reach the end of list 1, the rest of values in
                # the list 1 must also shows in list 1 only, we need to add them
                # to answer list too.
                values_list.append(list_1[i])

                i += 1

            return values_list

        answer_list.append(getDifferenceBetweenTwoIntListsValues(nums1, nums2))
        answer_list.append(getDifferenceBetweenTwoIntListsValues(nums2, nums1))

        return answer_list


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        [[1, 2, 3], [2, 4, 6]],
        [[1, 2, 3, 3], [1, 1, 2, 2]],
    ]

    for testcase in testcases_list:
        print(sol.findDifference(*testcase))

File: all/2215_find_the_difference_of_two_arrays/sol_sorting_deduplicate_inplace_two_pointers.c
#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>

typedef struct {
    int *array_ptr;
    int array_length;
    int next_available_index;
} DynamicIntArrayST;

static int compareFunction(const void *p1, const void *p2) {
    const int i1 = (*((const int*)p1));
    const int i2 = (*((const int*)p2));

    return ((i1 > i2) - (i1 < i2));
}

static bool addNewValueToDynamicIntArrayST(
    const int x,
    DynamicIntArrayST * dynamic_int_array_st_ptr
) {
    int *new_array_ptr = NULL;
    int new_array_length = 0;

    if (
        (dynamic_int_array_st_ptr->next_available_index)
        >=
        (dynamic_int_array_st_ptr->array_length)
    ) {
        if (0 == (dynamic_int_array_st_ptr->array_length)) {
            new_array_length = 4;
        }
        else {
            new_array_length = ((dynamic_int_array_st_ptr->array_length) << 1);
        }

        new_array_ptr = (
            (int*)realloc(
                dynamic_int_array_st_ptr->array_ptr,
                (
                    sizeof(*(dynamic_int_array_st_ptr->array_ptr))
                    *
                    new_array_length
                )
            )
        );

        if (NULL == new_array_ptr) {
            return false;
        }

        (dynamic_int_array_st_ptr->array_length) = new_array_length;
        (dynamic_int_array_st_ptr->array_ptr) = new_array_ptr;
    }

    (dynamic_int_array_st_ptr->array_ptr)[
        (dynamic_int_array_st_ptr->next_available_index)++
    ] = x;

    return true;
}

static int deduplicateSortedIntArray(
    int * const array_ptr,
    const int array_length
) {
    if (0 == array_length) {
        return 0;
    }

    int next_unique_element_index = 1;

    for (int i = 1; i < array_length; ++i) {
        if ((array_ptr[i - 1]) == (array_ptr[i])) {
            continue;
        }

        array_ptr[next_unique_element_index] = (array_ptr[i]);
        ++next_unique_element_index;
    }

    return next_unique_element_index;
}

static void getDifferenceBetweenTwoIntArraysValues(
    const int * const int_array_1_ptr,
    const int * const int_array_2_ptr,
    const int int_array_1_length,
    const int int_array_2_length,
    DynamicIntArrayST * dynamic_int_array_st_ptr
) {
    /*
     * Assume int array 1 and int array 2 are both sorted in the ascending
     * order and only have unique values.
     *
     * We use two pointers to find all unique values that are the int array 1
     * but not in the int array 2.
     */
    int i = 0, j = 0;

    while ((i < int_array_1_length) && (j < int_array_2_length)) {
        if ((int_array_1_ptr[i]) == (int_array_2_ptr[j])) {
            /*
             * As int_array_1_ptr[i] is equal to int_array_2_ptr[j], it cannot
             * be the value we want, just increase the index i and the index j
             * by 1.
             * The index j is also increased since all values in the int array 1
             * and the int array 2 are unique. There won't be other elements
             * whose value is int_array_1_ptr[i] in the rest of int array 1.
             */
            ++i;
            ++j;
        }
        else if ((int_array_1_ptr[i]) > (int_array_2_ptr[j])) {
            /*
             * As int_array_1_ptr[i] is greater than int_array_2_ptr[j],
             * we do not know whether int_array_1_ptr[i] only shows in the
             * int array 1, we need to increase the index j by 1 to check
             * more values.
             */
            ++j;
        }
        else {
            /*
             * As int_array_1_ptr[i] is less than int_array_2_ptr[j], and since
             * int array 2 is sorted, we cannot find int_array_1_ptr[i] in
             * the int_array_2_ptr[j:], so int_array_1_ptr[i] must only show in
             * the int array 1.
             * Also, all values in the given int array 1 are unique, we can add
             * int_array_1_ptr[i] to answer array.
             */
            addNewValueToDynamicIntArrayST(
                int_array_1_ptr[i],
                dynamic_int_array_st_ptr
            );

            ++i;
        }
    }

    while (i < int_array_1_length) {
        /*
         * If we still not reach the end of int arrary 1, the rest of values in
         * the int array 1 must also shows in int array 1 only, we need to add
         * them to answer array too.
         */
        addNewValueToDynamicIntArrayST(
            int_array_1_ptr[i],
            dynamic_int_array_st_ptr
        );

        ++i;
    }
}

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** findDifference(
    int* nums1,
    int nums1Size,
    int* nums2,
    int nums2Size,
    int* returnSize,
    int** returnColumnSizes
) {
    int **answer_array_ptr = (int**)malloc(sizeof(*answer_array_ptr) * 2);
    (*returnColumnSizes) = (int*)malloc(sizeof(**returnColumnSizes) * 2);

    int nums1_array_length_after_deduplicate = 0;
    int nums2_array_length_after_deduplicate = 0;
    DynamicIntArrayST nums1_answer_array_st = {
        .array_ptr = NULL,
        .array_length = 0,
        .next_available_index = 0,
    };
    DynamicIntArrayST nums2_answer_array_st = {
        .array_ptr = NULL,
        .array_length = 0,
        .next_available_index = 0,
    };

    /* Sort the nums1 array in the ascending order. */
    qsort(nums1, nums1Size, sizeof(*nums1), &compareFunction);

    /* Sort the nums2 array in the ascending order. */
    qsort(nums2, nums2Size, sizeof(*nums2), &compareFunction);

    nums1_array_length_after_deduplicate = deduplicateSortedIntArray(
        nums1,
        nums1Size
    );
    nums2_array_length_after_deduplicate = deduplicateSortedIntArray(
        nums2,
        nums2Size
    );

    /*
     * Get all the unique values that are in nums1 array but no in the nums2
     * array.
     */
    getDifferenceBetweenTwoIntArraysValues(
        nums1,
        nums2,
        nums1_array_length_after_deduplicate,
        nums2_array_length_after_deduplicate,
        &nums1_answer_array_st
    );

    /*
     * Get all the unique values that are in nums2 array but no in the nums1
     * array.
     */
    getDifferenceBetweenTwoIntArraysValues(
        nums2,
        nums1,
        nums2_array_length_after_deduplicate,
        nums1_array_length_after_deduplicate,
        &nums2_answer_array_st
    );

    answer_array_ptr[0] = (nums1_answer_array_st.array_ptr);
    (*returnColumnSizes)[0] = (nums1_answer_array_st.next_available_index);

    answer_array_ptr[1] = (nums2_answer_array_st.array_ptr);
    (*returnColumnSizes)[1] = (nums2_answer_array_st.next_available_index);

    (*returnSize) = 2;

    return answer_array_ptr;
}
